packages/shared-ux/table_persist/src/use_table_persist.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Criteria } from '@elastic/eui';
+import { PropertySort } from '@elastic/eui/src/services/sort/property_sort';
+import {
+  DEFAULT_INITIAL_PAGE_SIZE,
+  LOCAL_STORAGE_PREFIX,
+  DEFAULT_PAGE_SIZE_OPTIONS,
+} from './constants';
+import { createStorage } from './storage';
+import { validatePersistData } from './validate_persist_data';
+import { PersistData } from './types';
+
+interface EuiTablePersistProps<T> {
+  /** A unique id that will be included in the local storage variable for this table. */
+  tableId: string;
+  /** (Optional) Specifies a custom onTableChange handler. */
+  customOnTableChange?: (change: Criteria<T>) => void;
+  /** (Optional) Specifies a custom initial table sorting. */
+  initialSort?: PropertySort;
+  /** (Optional) Specifies a custom initial page size for the table.
+   * Defaults to {@link DEFAULT_INITIAL_PAGE_SIZE} */
+  initialPageSize?: number;
+  /** (Optional) Specifies custom page size options for the table.
+   * Defaults to {@link DEFAULT_PAGE_SIZE_OPTIONS} */
+  pageSizeOptions?: number[];
+}
+
+/**
+ * A hook that stores and retrieves from local storage the table page size and sort criteria.
+ * Returns the persisting page size and sort and the onTableChange handler that should be passed
+ * as props to an Eui table component.
+ */
+export const useEuiTablePersist = <T extends object>({
+  tableId,
+  customOnTableChange,
+  initialSort,
+  initialPageSize,
+  pageSizeOptions,
+}: EuiTablePersistProps<T>) => {
+  const storage = createStorage({
+    engine: window.localStorage,
```

Comment:
I think this level of abstraction from localStorage is needed if the consumer of this package can provide a storage engine themselves. Otherwise if the storage engine is hard coded, I think the engine might as well be hidden in the storage class.

Reply:
I think we could use a `useCallback` to wrap `onTableChange` that is returned to the consumer.

We could also have `sorting` and `pageSize` that are returned to the consumer as 2 states (their values would be updated in `onTableChange`) so they can react to their value change if need be?

Reply:
Good calls! Removed the engine abstraction from the Storage class and added `useCallback` and states with [096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b).